File: internal/models/tasks.go
package models

import (
	"github.com/volatiletech/null/v8"
)

type Task struct {
	ID                    uint               `json:"id" gorm:"Column:id"`
	Type                  int8               `json:"type" gorm:"Column:type"`
	Duration              uint               `json:"duration" gorm:"Column:duration"`
	DurationCode          int8               `json:"duration_code" gorm:"Column:duration_code"`
	Title                 string             `json:"title" gorm:"Column:title"`
	Description           string             `json:"description" gorm:"Column:description"`
	RewardText            string             `json:"reward_text" gorm:"Column:reward_text"`
	RewardIdList          null.String        `json:"reward_id_list" gorm:"Column:reward_id_list"`
	CashReward            uint               `json:"cash_reward" gorm:"Column:cash_reward"`
	ExpReward             int                `json:"exp_reward" gorm:"Column:exp_reward"`
	RewardMethod          uint8              `json:"reward_method" gorm:"Column:reward_method"`
	RewardPoints          int                `json:"reward_points" gorm:"Column:reward_points"`
	RewardPointType       int                `json:"reward_point_type" gorm:"Column:reward_point_type"`
	MinLevel              uint8              `json:"min_level" gorm:"Column:min_level"`
	MaxLevel              uint8              `json:"max_level" gorm:"Column:max_level"`
	LevelSpread           uint               `json:"level_spread" gorm:"Column:level_spread"`
	MinPlayers            uint               `json:"min_players" gorm:"Column:min_players"`
	MaxPlayers            uint               `json:"max_players" gorm:"Column:max_players"`
	Repeatable            uint8              `json:"repeatable" gorm:"Column:repeatable"`
	FactionReward         int                `json:"faction_reward" gorm:"Column:faction_reward"`
	CompletionEmote       string             `json:"completion_emote" gorm:"Column:completion_emote"`
	ReplayTimerGroup      uint               `json:"replay_timer_group" gorm:"Column:replay_timer_group"`
	ReplayTimerSeconds    uint               `json:"replay_timer_seconds" gorm:"Column:replay_timer_seconds"`
	RequestTimerGroup     uint               `json:"request_timer_group" gorm:"Column:request_timer_group"`
	RequestTimerSeconds   uint               `json:"request_timer_seconds" gorm:"Column:request_timer_seconds"`
	DzTemplateId          uint               `json:"dz_template_id" gorm:"Column:dz_template_id"`
	LockActivityId        int                `json:"lock_activity_id" gorm:"Column:lock_activity_id"`
	FactionAmount         int                `json:"faction_amount" gorm:"Column:faction_amount"`
	TaskActivities        []TaskActivity     `json:"task_activities,omitempty" gorm:"foreignKey:taskid;references:id"`
	Tasksets              []Taskset          `json:"tasksets,omitempty" gorm:"foreignKey:taskid;references:id"`
	AlternateCurrency     *AlternateCurrency `json:"alternate_currency,omitempty" gorm:"foreignKey:reward_point_type;references:id"`
}

func (Task) TableName() string {
    return "tasks"
}

func (Task) Relationships() []string {
    return []string{
		"AlternateCurrency",
		"AlternateCurrency.Item",
		"AlternateCurrency.Item.AlternateCurrencies",
		"AlternateCurrency.Item.CharacterCorpseItems",
		"AlternateCurrency.Item.DiscoveredItems",
		"AlternateCurrency.Item.Doors",
		"AlternateCurrency.Item.Doors.Item",
		"AlternateCurrency.Item.Fishings",
		"AlternateCurrency.Item.Fishings.Item",
		"AlternateCurrency.Item.Fishings.NpcType",
		"AlternateCurrency.Item.Fishings.NpcType.AlternateCurrency",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.LoottableEntries",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.LoottableEntries.Lootdrop",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.LoottableEntries.Lootdrop.LootdropEntries",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.LoottableEntries.Lootdrop.LootdropEntries.Item",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.LoottableEntries.Lootdrop.LootdropEntries.Lootdrop",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.LoottableEntries.Lootdrop.LoottableEntries",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.LoottableEntries.Loottable",
		"AlternateCurrency.Item.Fishings.NpcType.Loottable.NpcTypes",
		"AlternateCurrency.Item.Fishings.NpcType.Merchantlists",
		"AlternateCurrency.Item.Fishings.NpcType.Merchantlists.Items",
		"AlternateCurrency.Item.Fishings.NpcType.Merchantlists.NpcTypes",
		"AlternateCurrency.Item.Fishings.NpcType.NpcEmotes",
		"AlternateCurrency.Item.Fishings.NpcType.NpcFactions",
		"AlternateCurrency.Item.Fishings.NpcType.NpcFactions.NpcFactionEntries",
		"AlternateCurrency.Item.Fishings.NpcType.NpcFactions.NpcFactionEntries.FactionList",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpell",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.Aura",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.Aura.SpellsNew",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.BlockedSpells",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.Damageshieldtypes",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.Items",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.NpcSpellsEntries",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.SpellBuckets",
		"AlternateCurrency.Item.Fishings.NpcType.NpcSpell.NpcSpellsEntries.SpellsNew.SpellGlobals",
		"AlternateCurrency.Item.Fishings.NpcType.NpcTypesTint",
		"AlternateCurrency.Item.Fishings.NpcType.Spawnentries",
		"AlternateCurrency.Item.Fishings.NpcType.Spawnentries.NpcType",
		"AlternateCurrency.Item.Fishings.NpcType.Spawnentries.Spawngroup",
		"AlternateCurrency.Item.Fishings.NpcType.Spawnentries.Spawngroup.Spawn2",
		"AlternateCurrency.Item.Fishings.NpcType.Spawnentries.Spawngroup.Spawn2.Spawnentries",
		"AlternateCurrency.Item.Fishings.NpcType.Spawnentries.Spawngroup.Spawn2.Spawngroup",
		"AlternateCurrency.Item.Fishings.Zone",
		"AlternateCurrency.Item.Forages",
		"AlternateCurrency.Item.Forages.Item",
		"AlternateCurrency.Item.Forages.Zone",
		"AlternateCurrency.Item.GroundSpawns",
		"AlternateCurrency.Item.GroundSpawns.Zone",
		"AlternateCurrency.Item.ItemTicks",
		"AlternateCurrency.Item.Keyrings",
		"AlternateCurrency.Item.LootdropEntries",
		"AlternateCurrency.Item.LootdropEntries.Item",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LootdropEntries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Lootdrop",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.LoottableEntries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.AlternateCurrency",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Loottable",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Merchantlists",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Merchantlists.Items",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Merchantlists.NpcTypes",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcEmotes",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcFactions",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcFactions.NpcFactionEntries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcFactions.NpcFactionEntries.FactionList",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpell",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Aura",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Aura.SpellsNew",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.BlockedSpells",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Damageshieldtypes",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Items",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.NpcSpellsEntries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.SpellBuckets",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.SpellGlobals",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.NpcTypesTint",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Spawnentries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Spawnentries.NpcType",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Spawnentries.Spawngroup",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Spawnentries.Spawngroup.Spawn2",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Spawnentries.Spawngroup.Spawn2.Spawnentries",
		"AlternateCurrency.Item.LootdropEntries.Lootdrop.LoottableEntries.Loottable.NpcTypes.Spawnentries.Spawngroup.Spawn2.Spawngroup",
		"AlternateCurrency.Item.Merchantlists",
		"AlternateCurrency.Item.Merchantlists.Items",
		"AlternateCurrency.Item.Merchantlists.NpcTypes",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.AlternateCurrency",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.LoottableEntries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.LoottableEntries.Lootdrop",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.LoottableEntries.Lootdrop.LootdropEntries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.LoottableEntries.Lootdrop.LootdropEntries.Item",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.LoottableEntries.Lootdrop.LootdropEntries.Lootdrop",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.LoottableEntries.Lootdrop.LoottableEntries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.LoottableEntries.Loottable",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Loottable.NpcTypes",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Merchantlists",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcEmotes",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcFactions",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcFactions.NpcFactionEntries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcFactions.NpcFactionEntries.FactionList",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpell",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Aura",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Aura.SpellsNew",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.BlockedSpells",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Damageshieldtypes",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.Items",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.NpcSpellsEntries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.SpellBuckets",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcSpell.NpcSpellsEntries.SpellsNew.SpellGlobals",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.NpcTypesTint",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Spawnentries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Spawnentries.NpcType",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Spawnentries.Spawngroup",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Spawnentries.Spawngroup.Spawn2",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Spawnentries.Spawngroup.Spawn2.Spawnentries",
		"AlternateCurrency.Item.Merchantlists.NpcTypes.Spawnentries.Spawngroup.Spawn2.Spawngroup",
		"AlternateCurrency.Item.ObjectContents",
		"AlternateCurrency.Item.Objects",
		"AlternateCurrency.Item.Objects.Item",
		"AlternateCurrency.Item.Objects.Zone",
		"AlternateCurrency.Item.StartingItems",
		"AlternateCurrency.Item.StartingItems.Item",
		"AlternateCurrency.Item.StartingItems.Zone",
		"AlternateCurrency.Item.TradeskillRecipeEntries",
		"AlternateCurrency.Item.TradeskillRecipeEntries.TradeskillRecipe",
		"AlternateCurrency.Item.TributeLevels",
		"TaskActivities",
		"Tasksets",
	}
}

func (Task) Connection() string {
    return "eqemu_content"
}

File: frontend/src/app/tasks.ts
import {TaskActivityApi, TaskApi} from "@/app/api";
import {SpireApiClient} from "@/app/api/spire-api-client";
import {TASK_ACTIVITY_TYPE} from "@/app/constants/eq-task-constants";
import {HttpStatus} from "@/app/api/http-status";
import {SpireQueryBuilder} from "@/app/api/spire-query-builder";

export class Tasks {

  public static async getTasks() {
    const r = await (new TaskApi(SpireApiClient.getOpenApiConfig())).listTasks()
    if (r.status === HttpStatus.OK) {
      return r.data
    }

    return []
  }

  public static getTaskApi() {
    return (new TaskApi(SpireApiClient.getOpenApiConfig()));
  }

  public static getTaskActivitiesApi() {
    return (new TaskActivityApi(SpireApiClient.getOpenApiConfig()));
  }

  // TODO: bubble up error handling
  public static async getTask(taskId: number) {
    let request = (new SpireQueryBuilder())
      .includes(this.getRelationships())
      .get()

    // @ts-ignore
    request.id = taskId
    // @ts-ignore
    const r    = await this.getTaskApi().getTask(request)

    if (r.status === HttpStatus.OK) {
      let task = r.data

      // @ts-ignore
      // sort activityids before return
      if (task.task_activities) {
        // @ts-ignore
        task.task_activities.sort((a, b) => (a.activityid > b.activityid) ? 1 : -1)
      }

      return task
    }

    return {}
  }

  // TODO: bubble up error handling
  public static async getTaskWithActivities(taskId: number) {
    let request = (new SpireQueryBuilder())
      .includes(this.getRelationships())
      .get()

    // @ts-ignore
    request.id = taskId
    // @ts-ignore

    const r = await this.getTaskApi().getTask(request)

    if (r.status === HttpStatus.OK) {
      return r.data
    }

    return {}
  }

  // TODO: bubble up error handling
  public static async createTask(task: any) {
    // @ts-ignore
    return await this.getTaskApi().createTask({task: task}, {})
  }

  // TODO: bubble up error handling
  public static async updateTask(task: any) {
    // @ts-ignore
    return await this.getTaskApi().updateTask({id: task.id, task: task})
  }

  // TODO: bubble up error handling
  public static async deleteTaskWithActivities(task: any) {
    try {
      const r = await this.getTaskApi().deleteTask({id: task.id}, {})
      if (r.status === HttpStatus.OK) {
        await this.getTaskActivitiesApi().deleteTaskActivity({id: task.id}, {})
      }
    } catch (err) {
      if (err.response.data.error) {
        return err.response.data.error
      }
    }

    return ""
  }

  public static async createNewTaskActivity(task: any) {
    let taskActivity        = this.getExampleActivity()
    taskActivity.taskid     = task.id
    taskActivity.activityid = (task.task_activities ? this.getNextActivityId(task.task_activities) : 0)
    taskActivity.step       = (task.task_activities ? this.getLatestStep(task.task_activities) : 1)

    // @ts-ignore
    return await this.getTaskActivitiesApi()
      .createTaskActivity(
        {
          taskActivity: taskActivity
        }
      )
  }

  public static async updateTaskActivityId(taskActivity: any, previousId: number) {
    let request = (new SpireQueryBuilder())
      .where("taskid", "=", taskActivity.taskid)
      .where("activityid", "=", previousId)
      .get()

    // format zones
    taskActivity.zones = taskActivity.zones.toString()

    // @ts-ignore
    return await this.getTaskActivitiesApi()
      .updateTaskActivity(
        {
          id: taskActivity.taskid,
          taskActivity: taskActivity
        },
        {query: request}
      )
  }

  public static async cloneTaskActivity(task: any, sourceActivityId) {
    let taskActivity        = task.task_activities[sourceActivityId]
    taskActivity.taskid     = task.id
    taskActivity.activityid = (task.task_activities ? this.getNextActivityId(task.task_activities) : 0)
    taskActivity.step       = (task.task_activities ? this.getLatestStep(task.task_activities) : 1)

    // @ts-ignore
    return await this.getTaskActivitiesApi()
      .createTaskActivity(
        {
          taskActivity: taskActivity
        }
      )
  }

  // TODO: bubble up error handling
  public static async deleteTaskActivity(activity: any) {
    let request = (new SpireQueryBuilder())
      .where("activityid", "=", activity.activityid)
      .get()

    // @ts-ignore
    request.id = activity.taskid

    // @ts-ignore
    return await this.getTaskActivitiesApi()
      .deleteTaskActivity(
        {id: activity.taskid},
        {query: request}
      )
  }

  private static getNextActivityId(activities: any) {
    let nextId = 0
    activities.forEach((a) => {
      nextId = a.activityid
    })

    return nextId + 1
  }

  private static getLatestStep(activities: any) {
    let step = 0
    activities.forEach((a) => {
      step = a.step
    })

    return step
  }

  public static buildActivityDescription(activity: any) {
    if (activity.description_override !== "") {
      return activity.description_override;
    }

    switch (activity.activitytype) {
      case TASK_ACTIVITY_TYPE.DELIVER:
        return `Deliver ${activity.goalcount} ${activity.item_list} to ${activity.target_name}`;
      case TASK_ACTIVITY_TYPE.KILL:
        return `Kill ${activity.goalcount} ${activity.target_name}`;
      case TASK_ACTIVITY_TYPE.LOOT:
        return `Loot ${activity.goalcount} ${activity.item_list} from ${activity.target_name}`;
      case TASK_ACTIVITY_TYPE.SPEAK_WITH:
        return "Speak with " + activity.target_name;
      case TASK_ACTIVITY_TYPE.EXPLORE:
        return "Explore " + activity.target_name;
      case TASK_ACTIVITY_TYPE.TRADESKILL:
        return `Create ${activity.goalcount} ${activity.item_list} using tradeskills`
      case TASK_ACTIVITY_TYPE.FISH:
        return `Fish for ${activity.goalcount} ${activity.item_list}`;
      case TASK_ACTIVITY_TYPE.FORAGE:
        return `Forage ${activity.goalcount} ${activity.item_list}`;
      case TASK_ACTIVITY_TYPE.USE:
        return "Use " + activity.goalcount;
      case TASK_ACTIVITY_TYPE.USE2:
        return "Use " + activity.goalcount;
      case TASK_ACTIVITY_TYPE.TOUCH:
        return "Touch " + activity.target_name;
      case TASK_ACTIVITY_TYPE.GIVE:
        return "Give " + activity.goalcount + " to " + activity.target_name;
      case TASK_ACTIVITY_TYPE.QUEST_SCRIPT_DRIVEN:
        return "None";
      default:
        return "None";
    }
  }

  public static getRelationships() {
    return [
      "AlternateCurrency.Item",
      "TaskActivities",
      // "TaskActivities.Goallists",
      // "TaskActivities.NpcType",
      // "TaskActivities.NpcType.Spawnentries",
      // "TaskActivities.NpcType.Spawnentries.NpcType",
      // "TaskActivities.NpcType.Spawnentries.Spawngroup",
      // "TaskActivities.NpcType.Spawnentries.Spawngroup.Spawn2",
      // "Tasksets",
    ]
  }

  public static getFieldDescriptions() {
    return {
      "id": "Task identifier",
      "step": "This is the logical step of your task activity, you can have many activities in one step, you have to complete all activities in one step to unlock the next step",
      "activitytype": "This is the type of task activity, kill, loot etc.",
      "item_list": "Name of items mentioned in task description",
      "npc_match_list": "Names or IDs of NPC's separated by | Example (10343|orc|gnoll)",
      "item_id_list": "IDs of items separated by | Ex: (10343|24232|34243)",
      "target_name": "This describes what the activity is targeting, it is different depending on the activity. For example 'orcs' Would display 'Kill X orcs'",
      "optional": "Describes whether or not this activity is optional",
      "goalid": "Goal ID is different depending on your activity type as well as Goal Method",
      "goalmethod": "Determine if this activity is updated via Quest or not.",
      "goalcount": "Required count of this activity to be completed.",
      "zones": "Zone ID(s) that correspond to this update. For example the client needs to be in this zone in order to receive the update. [-1 = Unknown, 0 = ALL] | Multiple example (3,10,30)",
      "zone_version": "Zone version(s) that correspond to this update. (-1 = ALL)",
      "description_override": "Use this to completely override other fields that otherwise influence the activity description (activity target, item list etc.)",
      "title": "The title of your task",
      "type": "The type of task you are making",
      "description": "Narrative description of your task",
      "duration_code": "Reflects the type of duration",
      "duration": "Task duration in seconds",
      "min_level": "Minimum level to receive the task",
      "max_level": "Maximum level to receive the task",
      "completion_emote": "Emote text that is displayed upon succesful completion of the task",
      "reward_method": "Determine if reward is quest controlled or not(s)",
      "reward_text": "Name of reward provided for completing the task. If this is empty, it is automatically derived from item name",
      "reward_id_list": "Item ID(s) for the rewarded item(s) separated by | Ex: (1001|1002|1003)",
      "exp_reward": "Amount of experience rewarded",
      "faction_reward": "Amount of faction rewarded",
      "reward_ebon_crystals": "Number of ebon crystals rewarded",
      "reward_radiant_crystals": "Number of radiant crystals rewarded",
      "cash_reward": "Amount of coin rewarded in copper",
    }
  }

  public static getFieldDescription(field: string) {
    // we do this because the payload we get back from spire API is
    // formatted slightly different
    let fieldLookup = field.toLowerCase().replace("_", "")

    for (let key in this.getFieldDescriptions()) {
      let keyLookup = key.toLowerCase().replace("_", "")
      if (keyLookup === fieldLookup) {
        return this.getFieldDescriptions()[key]
      }
    }

    return ''
  }

  // used to seed "new" tasks in the task editor
  public static getExampleTask() {
    return {
      "id": 2,
      "type": 2,
      "duration": 3600,
      "duration_code": 0,
      "title": "Example Task 2",
      "description": "Example Task 2",
      "reward_text": "XP",
      "reward_id_list": 0,
      "cash_reward": 0,
      "exp_reward": 10,
      "reward_method": 0,
      "lock_activity_id": -1,
      "reward_point_type": 0,
      "reward_points": 0,
      "min_level": 0,
      "max_level": 0,
      "level_spread": 0,
      "min_players": 0,
      "max_players": 0,
      "repeatable": 1,
      "faction_reward": 0,
      "completion_emote": "",
      "dz_template_id": 0,
      "replay_timer_seconds": 0,
      "replay_timer_group": 0,
      "request_timer_seconds": 0,
      "request_timer_group": 0,
      "task_activities": [{
        "taskid": 2,
        "activityid": 0,
        "step": 1,
        "activitytype": 2,
        "req_activity_id": -1,
        "target_name": "Orcs",
        "item_list": "",
        "item_id_list": "",
        "npc_match_list": "",
        "skill_list": "-1",
        "spell_list": "0",
        "description_override": "",
        "goalid": 0,
        "goalmethod": 0,
        "goalcount": 1,
        "zones": "21",
        "zone_version": -1,
        "optional": 0
      }, {
        "taskid": 2,
        "activityid": 1,
        "req_activity_id": -1,
        "step": 1,
        "activitytype": 3,
        "target_name": "any creature",
        "item_list": "Rusty Items",
        "item_id_list": "",
        "npc_match_list": "",
        "skill_list": "-1",
        "spell_list": "0",
        "description_override": "",
        "goalid": 0,
        "goalmethod": 0,
        "goalcount": 1,
        "zones": "0",
        "zone_version": -1,
        "optional": 0
      }, {
        "taskid": 2,
        "activityid": 2,
        "req_activity_id": -1,
        "step": 1,
        "activitytype": 5,
        "target_name": "",
        "item_list": "",
        "item_id_list": "",
        "npc_match_list": "",
        "skill_list": "-1",
        "spell_list": "0",
        "description_override": "Locate the Antonica Spires in the Luclin Nexus",
        "goalid": 0,
        "goalmethod": 0,
        "goalcount": 1,
        "zones": "152",
        "zone_version": -1,
        "optional": 0
      }]
    }
  }

  public static getExampleActivity() {
    return {
      "taskid": 2,
      "activityid": 0,
      "step": 1,
      "activitytype": 2,
      "req_activity_id": -1,
      "target_name": "Orcs",
      "item_list": "",
      "item_id_list": "",
      "npc_match_list": "",
      "skill_list": "-1",
      "spell_list": "0",
      "description_override": "",
      "goalmethod": 0,
      "goalcount": 1,
      "delivertonpc": 0,
      "zones": "21",
      "zone_version": -1,
      "optional": 0
    }
  }

}
